### PR TITLE
A ClowderConfigSource for Quarkus

### PR DESCRIPTION
**Pull request by pilhuhn**
This allows to configure Quarkus via the cdappconfig.json file in a transparent manner. 
(Almost) no code changes should be needed in Quarkus.
To use pull this dependency in as maven package and add a service-loader file.

Relates to [NOTIF-146]

cc @bennetelli

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+[source,xml]
+----
+<dependency>
+  <groupId>com.redhat.cloud.common</groupId>
+  <artifactId>clowder-quarkus-config-source</artifactId>
+  <version>0.1.0-SNAPSHOT</version>
+</dependency>
+----
+
+Then add the config file for the service loader at `META-INF/services/io.smallrye.config.ConfigSourceFactory`:
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply README.adoc:17]: I think I have seen some oddities when I had it in here and not in the depending project. I forgot. Probably need to revisit.

File: src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    @Override
+    public Set<String> getPropertyNames() {
+        return existingValues.keySet();
+    }
+
+    @Override
+    public int getOrdinal() {
+        // Provide a value higher than 250 to it overrides application.properties
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java:67]: Yes. Also using this ConfigSource(Factory) is pretty special so I don't expect us to put a wild-west of other sources next to it.

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            <configuration>
+              <tagNameFormat>v@{project.version}</tagNameFormat>
+            </configuration>
+          </plugin>
+
+        </plugins>
+      </build>
+    </profile>
+    <profile>
+      <id>native</id>
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply pom.xml:180]: no

File: src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+        if (configKey.equals("quarkus.http.port")) {
+            JsonNumber webPort = root.getJsonNumber("webPort");
+            return webPort.toString();
+        }
+        if (configKey.equals("kafka.bootstrap.servers")) {
+            JsonArray brokers = root.getJsonObject("kafka").getJsonArray("brokers");
+            JsonObject broker = brokers.getJsonObject(0);
+            String b = broker.getString("hostname") + ":" + broker.getJsonNumber("port").toString();
+            return b;
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java:95]: I guess so. :) Not sure I have seen that yet over here, but it is better to prepare for it.

File: src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    @Override
+    public String getValue(String configKey) {
+
+        // This matches against the property as in application.properties
+        // For profiles != prod, values are requested first like
+        // %<profile>.property. E.g. %dev.quarkus.http.port
+
+
+        if (configKey.equals("quarkus.http.port")) {
+            JsonNumber webPort = root.getJsonNumber("webPort");
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java:88]: I need to do the getJsonNumber() as those are represented as numbers in Json. getString() throws as ClassCastException.

File: README.adoc [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+|false
+
+|clowder.file
+|/cdappconfig/cdappconfig.json
+
+|===
+
+.NOTE
+For Quarkus puting this in `application.properties` currently does not work.
+The flags need to be given via system property (`-Dclowder.use-source=true`) o environment (`CLOWDER_USE_SOURCE=TRUE java  -jar ...`)
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply README.adoc]: This is solved meanwhile

[the remark this answers is not in the document]
Comment (gwenneg) [reply README.adoc]: Great!

File: src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwenneg) [reply src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java]: So it means a ClassCastException will be thrown if we have something like `"webPort": "8000"`.

Can that happen and if so, should we throw something cleaner that would help fixing the issue?

[the remark this answers is not in the document]
Comment (pilhuhn) [reply src/main/java/com/redhat/cloud/common/clowder/configsource/ClowderConfigSource.java]: I think I am assuming that the Clowder-team does the right thing. The file I test against is a real one, where some entries are obfuscated.
I think the good thing is that if the CCE is thrown, it is pretty obvious and we can still fix it then.
cc @bsquiz @psav

**Review by gwenneg — APPROVED** [review]
LGTM

**Comment by bennetelli** [comment]
looks like I'm too late to the party..

**Pull request merged**